### PR TITLE
Add view guidelines section to CONTRIBUTING.md

### PR DESCRIPTION
**Pull request by seadowg**
Motivated by [this write up](https://gist.github.com/seadowg/93f33419460bcc7895990dab5c5958c5) of pains experienced during a refactor.

Interested to see if people agree with the guidelines!

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -119,6 +121,16 @@ text color | themeUtils.getPrimaryTextColor() | ?primaryTextColor
 accent color | themeUtils.getAccentColor() | ?colorAccent
 icon color | themeUtils.getIconColor() | ?iconColor
 
+## Custom view guidelines
+
+When creating or refactoring views, keep in mind our vision of an "ideal" view:
+
+- Views should generally be as dumb and as stateless as possible
+- View's shouldn't interact with other layers of your application (repositories, network etc). They should be able to take in and render data (via setters) and emit "meaningful" events via listeners.
```
Comment (lognaturel) [inline CONTRIBUTING.md]: "View's" -> "Views"

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -119,6 +121,16 @@ text color | themeUtils.getPrimaryTextColor() | ?primaryTextColor
...
+- A view should have a single public `Listener` sub interface and `setListener` method. Methods on the interface should correspond to "meaningful" events. If the view is a button that could just be `onClick` but if it's a volume slider this might be something like `onVolumeChanged`.
+- Views would ideally have just one setter for data but more complex views (often that have many subviews) that take in data at different times may have more - less setters is better as it's less changing state in the view.
```
Comment (lognaturel) [inline CONTRIBUTING.md]: total nitpick but "less setters" -> "fewer setters"

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -119,6 +121,16 @@ text color | themeUtils.getPrimaryTextColor() | ?primaryTextColor
...
+- Views that render more than one kind of data (and that have more than one setter) might benefit from `render` method that encapsulates all the logic around displaying the state of a view.
```
Comment (lognaturel) [inline CONTRIBUTING.md]: "benefit from" -> "benefit from a"

**Review by lognaturel — CHANGES_REQUESTED** [review]

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
@seadowg can you please resolve conflicts and we can merge?

**Review by grzesiek2010 — APPROVED** [review]

**Pull request merged**